Issue 562  typset math in proximals

aringh opened the pull request:
This is not completely done yet, but I'm creating the pull request so that @kohr-h and @adler-j can start checking if it looks ok.

As the name says, this pull request is mainly #562, but as I was anyway working with the proximals I also incorporated in this branch #572.

adler-j commented:
The name change really should be cherry picked to another branch and a separate pull request, that would make this easier to review.

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -524,9 +526,36 @@ def proximal_nonnegativity(space):
     prox_factory : function
         Factory for the proximal operator to be initialized
 
+    Notes
```
This may be redundant, perhaps simply reference the box constraint.

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -568,6 +588,33 @@ def proximal_cconj_l2(space, lam=1, g=None):
 
     Notes
     -----
+    The functional :math:`F` is given by
```
The l2-norm functional ...

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -568,6 +588,33 @@ def proximal_cconj_l2(space, lam=1, g=None):
+        .. math::
+
+            prox[\\sigma F^*](y) = \\left\{ \\begin{array}{ll}
+            \\lambda \\frac{y - g}{||y - g||}
```
Where did sigma go?

aringh replied:
That's a very good question... I just typeset what it said before and there is no `sigma` there. But I will double check.

aringh replied:
I don't think so... but I will add a comment about it.

kohr-h replied:
I think in all examples here, you just get the product `lambda * sigma`, so in some way it's a bit unnecessary to carry through both lambda and sigma.

aringh replied:
`lambda` has an effect, since that decides the region for the indicator function. But `sigma` do not mattes, since (positive) scaling the indicator function is still the indicator function.

kohr-h replied:
You're right, it makes a difference if we're looking at the proximals of convex conjugates. But for the non-cconj ones, lambda and sigma play the same roles.

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -706,6 +755,30 @@ def proximal_cconj_l2_squared(space, lam=1, g=None):
+        .. math::
+
+            F^*(y) = \\frac{1}{\\lambda} \left( \left|\left|
+            \\frac{y}{\\lambda} \\right| \\right|_2^2 +
```
Shouldnt there be a `1/4` somewhere here?

aringh replied:
Yes, there should. Again, I was only typesetting what it already said but I think this expression is wrong... my hand calculations give `1/(4lam) ||y||_2 ^2 + <y,g>`, which also rings a bell.

aringh replied:
But the expression for the proximal was still correct (which is what I hoped/expected, since it is the same as implemented and it passes the tests)

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -846,6 +898,61 @@ def proximal_cconj_l1(space, lam=1, g=None, isotropic=False):
+            \left\langle \\frac{y}{\\lambda}, g \\right\\rangle
+            \\right)
+
+    For a step size ``sigma``, the proximal operator of ``sigma * F^*`` is
```
Typeset

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -568,6 +588,33 @@ def proximal_cconj_l2(space, lam=1, g=None):
+        .. math::
+
+            F^*(y) = \\left\{ \\begin{array}{ll}
+            0 & \\quad  ||y-g||_2 \leq \\lambda, \\\\
```
Shouldnt it be something like ``\leq 1/(4 lambda``?

aringh replied:
I don't think so... page 10 and 11 [here](http://www.seas.ucla.edu/~vandenbe/236C/lectures/conj.pdf) says that should not be the case. I think it is for `|| . ||_2^2` you have it.

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -613,6 +652,25 @@ def proximal_l2(space, lam=1, g=None):
+            g & \\text{else},
+            \\end{array} \\right.
+
+    where :math:`c = \\sigma \cdot \\frac{\\lambda}{||y - g||_2}`.
```
remove the cdots, they may be confused with dot products

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -613,6 +652,25 @@ def proximal_l2(space, lam=1, g=None):
+            prox[\\sigma F](y) = \\left\{ \\begin{array}{ll}
+            \\frac{1 - c}{||y-g||} \\cdot y  + c \cdot g
+            & \\quad  \\text{if } c < g, \\\\
+            g & \\text{else},
```
You have a quad before the if statement, but not the else, so this looks weird when typeset.

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -776,12 +844,27 @@ def proximal_l2_squared(space, lam=1, g=None):
+
+        .. math::
+
+            prox[\\sigma F](x) = \\frac{x + 2 * \\sigma * \\lambda * g}
```
Remove the *, doesnt typeset well in math.

adler-j commented:
### Warnings when building

General warnings from building the docs:

```
/home/jonasadler/Programming/ODL/odl/solvers/functional/default_functionals.py:docstring of odl.solvers.functional.default_functionals.IndicatorLpUnitBall.proximal:6: WARNING: Title underline too short.

See Also
-----
/home/jonasadler/Programming/ODL/odl/solvers/functional/default_functionals.py:docstring of odl.solvers.functional.default_functionals.IndicatorLpUnitBall.proximal:6: SEVERE: Unexpected section title.

See Also
-----
/home/jonasadler/Programming/ODL/odl/solvers/functional/default_functionals.py:docstring of odl.solvers.functional.default_functionals.L1Norm.proximal:6: WARNING: Title underline too short.

See Also
-----
/home/jonasadler/Programming/ODL/odl/solvers/functional/default_functionals.py:docstring of odl.solvers.functional.default_functionals.L1Norm.proximal:6: SEVERE: Unexpected section title.

See Also
-----
/home/jonasadler/Programming/ODL/odl/solvers/functional/default_functionals.py:docstring of odl.solvers.functional.default_functionals.L2Norm.proximal:6: WARNING: Title underline too short.

See Also
-----
/home/jonasadler/Programming/ODL/odl/solvers/functional/default_functionals.py:docstring of odl.solvers.functional.default_functionals.L2Norm.proximal:6: SEVERE: Unexpected section title.

See Also
-----
/home/jonasadler/Programming/ODL/odl/solvers/functional/default_functionals.py:docstring of odl.solvers.functional.default_functionals.L2NormSquared.proximal:6: WARNING: Title underline too short.

See Also
-----
/home/jonasadler/Programming/ODL/odl/solvers/functional/default_functionals.py:docstring of odl.solvers.functional.default_functionals.L2NormSquared.proximal:6: SEVERE: Unexpected section title.

See Also
-----

```

also

```
/home/jonasadler/Programming/ODL/doc/source/math/solvers/advanced/proximal_operators.rst:60: WARNING: citation not found: R1970
```

### Places that were not updated

Further, the following proximals do not seem to be changed to the new style

* proximal_arg_scaling
* proximal_cconj
* proximal_const_func
* proximal_composition
* proximal_quadratic_perturbation
* proximal_translation

### Prox in cursive

Also, prox is typeset *in cursive*, I think we want it in upright font.

### Norms -> Distances

You also call constructions like ``||x - g||`` norms, more accurately they would be distances.

### Typeset initial def as code

It seems we have a old problem with comments like like

```
This is calculated according to the rule

        prox[s * F( . - y)](x) = y + prox[s * F](x - y)
```

This does not typeset very well, and would look much better in case we added two colons,

```
This is calculated according to the rule::

        prox[s * F( . - y)](x) = y + prox[s * F](x - y)
```

### Distinction between elementwise and global definitions

We have an ambiguity in that we sometimes use `x` to refer to a single point in some `R^n` style space, and sometimes to the whole space. In the second case, we sometimes index `x` by `i`, and sometimes we dont.

For examples of the styles

* `x` is a point: `proximal_nonnegativity`
* `x` is a vector: `proximal_cconj_l2`
* `x` is a vector that we index: `proximal_cconj_kl_cross_entropy`

For this thing, I guess we simply need a discussion on how to handle this, or if we are fine with this (I guess it is obvious from reading what we intend?)

aringh commented:
As I wrote at the top, this is not done yet ;) But it was exactly comments like

> Prox in cursive
> [...]
> Norms -> Distances
> [...]
> Typeset initial def as code
> [...]
> Distinction between elementwise and global definitions

that I created the pull request anyway :) I will look at the comments, and also update the ones which have not been typeset yet.

Also: I will try to cherry pick the renaming commit and move elsewhere.

kohr-h commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -446,6 +436,33 @@ def proximal_box_constraint(space, lower=None, upper=None):
+    It is independent of :math:`\\sigma` and invariant under a positive
+    rescaling of :math:`I_{x \\in P}`, since that leaves the indicator function
+    unchanged.
+
```
Pointwise? Rather per axis or per component.

kohr-h commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -524,9 +526,36 @@ def proximal_nonnegativity(space):
+
+        .. math::
+
+            I_{x \\in P} = \\left\{ \\begin{array}{ll}
```
Rather `I_P(x)`, or `\\iota_P(x)` - the last notation seems to be rather wide spread for indicator functions.

kohr-h commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -446,6 +436,33 @@ def proximal_box_constraint(space, lower=None, upper=None):
+
+        .. math::
+
+            I_{x \\in P} = \\left\{ \\begin{array}{ll}
```
`I_P(x)` or `\\iota_P(x)`

kohr-h replied:
Specific reason not to use `\begin{cases} \end{cases}`? It doesn't require the `\left \right` construction and sets appropriate horizontal spaces automatically.

aringh replied:
Have never used it, was not aware if the existence of that functionality :)

kohr-h commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -446,6 +436,33 @@ def proximal_box_constraint(space, lower=None, upper=None):
+
+      .. math::
+
+             prox[\\sigma I_{x \\in P}](x) = \\left\{ \\begin{array}{ll}
```
I don't quite like our old style of `prox[functional](point)`, i.e. the functional in square parentheses. IMO it looks better as subscript and upright font, i.e. `\mathrm{prox}_{\\iota_P}(x)`. Should be fixable by simple search-and-replace.

kohr-h commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -568,6 +588,33 @@ def proximal_cconj_l2(space, lam=1, g=None):
+
+        .. math::
+
+            F(x) = \\lambda ||x - g||_2
```
Use `\| x - g \|` to typeset the norm. It doesn't impair the (lower priority) plain text readability significantly but improves the typeset formula.
Search-and-replace across the board.

kohr-h commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -568,6 +588,33 @@ def proximal_cconj_l2(space, lam=1, g=None):
+
+    For a step size :math:`\\sigma`, the proximal operator of
+    :math:`sigma F^*` is given by the projection onto the set
+    :math:`||y-g||_2 \leq \\lambda`, i.e., by
```
Strictly speaking, this is not a set. Simple fix: "onto the set of y satisfying ..."

aringh commented:
We have a naming issue about the step-size parameter:

 - `proximal_quadratic_perturbation` calls it `step_size`
- `proximal_const_func` calls it `tau`
- `proximal_l2` calls it `sigma`

Is this multitude of names ok, or should we harmonize? If so, to what?

kohr-h commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -846,6 +898,61 @@ def proximal_cconj_l1(space, lam=1, g=None, isotropic=False):
 
+    Notes
+    -----
+    The :math:`L_1`-norm :math:`F` is the functional
```
distance
Also I prefer 1 as exponent up there, but that's a matter of taste. We should be consistent though.

kohr-h commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -846,6 +898,61 @@ def proximal_cconj_l1(space, lam=1, g=None, isotropic=False):
+            F(x) = \\lambda ||x - g||_1.
+
+    The convex conjugate :math:`F^*` of :math:`F` is given by the indicator
+    function of the set :math:`box(\\lambda)`
```
This won't look pretty. I'd replace `box` by a letter and use `lambda` as subscript. The letter `Q` is frequently used for boxes, so that would be my choice.

kohr-h commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -846,6 +898,61 @@ def proximal_cconj_l1(space, lam=1, g=None, isotropic=False):
+            F^*(y) = \\lambda I_{box(\\lambda)} \left(
+            \left| \\frac{y}{\\lambda} \\right| +
+            \left\\langle \\frac{y}{\\lambda}, g \\right\\rangle
+            \\right)
```
I don't get this. We have a box scaled by `lambda`, and then again a scaling with `1/lambda` in the argument. That should be easy to combine, or even one of the two is redundant.

aringh replied:
I will first give the same answer as I did somewhere else: I have simply typeset this `F^*(y) = lam ind_{box(lam)}(|y / lam| + <y / lam, g>)`

To answer the real question: we have an indicator function scaled by `lambda`. That scaling is pointless I guess, since that function only takes value 0 or infinity anyway. But I don't think it is necessarily that easy to simplify the argument of the indicator function... you can maybe try some completion of squares, but that I don't think it is very useful.

kohr-h replied:
> I will first give the same answer as I did somewhere else: I have simply typeset this

Sure :-)



> But I don't think it is necessarily that easy to simplify the argument of the indicator function... you can maybe try some completion of squares, but that I don't think it is very useful.

I think it's quite trivial. You can extract a 1/lam from both summands.

My suspicion is that the formula is simply wrong since the scaling of the argument by `1/lam` is the same as scaling the box by `lam`. Probably it should be either 
```
box(1)(|y / lam| + <y / lam, g>)
```
or
```
box(lam)(|y| + <y, g>)
```

kohr-h commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -966,11 +1060,44 @@ def proximal_l1(space, lam=1, g=None, isotropic=False):
+            \\sigma \\lambda \\\\
+            y + \\sigma \\lambda
+            & \\quad  \\text{if } y < g - \\sigma \\lambda,
+            \\end{array} \\right.
```
This is a bit confusing since the inequalities in the cases have to read pointwise (I guess), i.e. per point or alternatively per vector entry.

kohr-h commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -966,11 +1060,44 @@ def proximal_l1(space, lam=1, g=None, isotropic=False):
+            F(x) = \\lambda || ||x - g||_2 ||_1
+
+    The proximal can be calculated using the Moreau equality (also known as
+    Moreau decomposition or Moreau idnetity). See for example [BC2011]_.
```
idnetity

kohr-h commented:
> Is this multitude of names ok, or should we harmonize? If so, to what?

No, this multitude is not okay. Since we figured that it's not a step size technically, we shouldn't call it that way. I'd go for `sigma`.

aringh commented:
All issues except the 

> Distinction between elementwise and global definitions

should have been addressed now. Do we want point-wise definitions or not?

adler-j commented:
> should have been addressed now. Do we want point-wise definitions or not?

I think we should simply be clear in what cases we have pointwise defintions and where we have global ones. The only ones I'd explicitly remove is where we make pointwise defintions, but dont use a subscript or anything to indicate that we are doing it.

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -117,24 +112,43 @@ def proximal_cconj(prox_factory):
         """
-        prox_other = (step_size * prox_factory(1.0 / step_size) *
-                      (1.0 / step_size))
+        prox_other = (sigma * prox_factory(1.0 / sigma) *
```
Perhaps fits on a line now?

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -342,26 +362,44 @@ def proximal_composition(proximal, operator, mu):
         Ir = IdentityOperator(operator.range)
-        prox_muf = proximal(step_size)
+        prox_muf = proximal(sigma)
         return Id + (1.0 / mu) * operator.adjoint((prox_muf - Ir) * operator)
```
There has to be a bug here, `operator.adjoint((prox_muf - Ir) * operator)` makes no sense, it should be `operator.adjoint * ((prox_muf - Ir) * operator)` I guess.

aringh replied:
You are probably right... and what is unfortunate is that we don't seem to have a test that covers this one.

The identity can be found in the table in [this article (Chp 10 by Combettes and Pesquet)](https://www.ljll.math.upmc.fr/~plc/prox.pdf), but they don't do it for arbitrary step lengths. So I guess one has to work that through manually...

Someone asked the same question on [math.stackexchange](http://math.stackexchange.com/questions/147987/proximal-mapping-for-composition-of-functions), and the main proof idea is given there.

I will try to work it throw.

aringh replied:
All statements I have found on this identity are without step length, and the proofs are of the form "applying lemma xxx, proposition yyy and proposition zzz the result follows directly", and hence hard to follow and modify.

On the other hand, let `f = gamma * g`. Then I guess the identity holds, for `f` and hence for any function `g` we get the sought identity with step length `gamma`. Does this sound reasonable?

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -387,14 +420,26 @@ def proximal_const_func(space):
+
+    Notes
+    -----
+    The constant functional :math:`G` is defeind as :math:`G(x) = constant`
```
defeind

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -568,8 +609,37 @@ def proximal_cconj_l2(space, lam=1, g=None):
+
+    Note that the expression is independent of :math:`\sigma`.
+
+    Most problems are forumlated for the squared norm/distance, in that case
```
This should likely come earlier.

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -966,11 +1082,44 @@ def proximal_l1(space, lam=1, g=None, isotropic=False):
+            y + \\sigma \\lambda
+            & \\text{if } y < g - \\sigma \\lambda,
+            \\end{cases}
+
```
One newline too much

aringh commented:
I have fixed the bug in `proximal_composition`, and also added a small test for it (not very fancy, but something). I see that this branch now have conflicts with master, and given the discussion in #595 I would prefer if someone could help me to re-base.

aringh commented:
Now it is also rebased. Good to go?

What could need some attanetion is the bug fix [here](https://github.com/odlgroup/odl/pull/580/files#diff-3c586797eacf683af348bca4457a86d2R403) and the test [here](https://github.com/odlgroup/odl/pull/580/files#diff-bb3fae52a9221094ec5adf2f6ed6ece3R171).

adler-j commented on test/solvers/advanced/proximal_utils_test.py:
```diff
@@ -168,5 +168,44 @@ def test_proximal_quadratic_perturbation_linear_and_quadratic():
     assert all_almost_equal(prox(x), expected_result, places=PLACES)
 
 
+def test_proximal_compositio():
```
def test_proximal_compositio**n**

adler-j commented on test/solvers/advanced/proximal_utils_test.py:
```diff
@@ -168,5 +168,44 @@ def test_proximal_quadratic_perturbation_linear_and_quadratic():
+    prox_factory = odls.proximal_l2_squared(space)
+
+    # The semi-orthogonal linear operator
+    scal = np.random.rand()
```
just pick a value

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -117,24 +112,42 @@ def proximal_cconj(prox_factory):
+    The Moreau identity states that for any convex function :math:`F` with
+    convex conjugate :math:`F^*`, the proximals satisfy
+
+        .. math::
```
one indentation step too deep

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -163,15 +170,24 @@ def proximal_translation(prox_factory, y):
     -----
+    Given a functional :math:`F`, this is calculated according to the rule
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -212,44 +221,45 @@ def proximal_arg_scaling(prox_factory, scaling):
     -----
+    Given a functional :math:`F`, this is calculated according to the rule
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -212,44 +221,45 @@ def proximal_arg_scaling(prox_factory, scaling):
+            \mathrm{prox}_{\\sigma F( \cdot scal)}(x) =
+            \\frac{1}{scal} \mathrm{prox}_{\\sigma \, scal^2 \, F }(x \, scal)
+
+    where ``scal`` is the scaling parameter, and :math:`\\sigma` is the step
```
typeset ``scal``

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -269,6 +279,25 @@ def proximal_quadratic_perturbation(prox_factory, a, u=None):
     -----
+    Given a functional :math:`F`, this is calculated according to the rule
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -269,6 +279,25 @@ def proximal_quadratic_perturbation(prox_factory, a, u=None):
+
+    where :math:`c` is the constant
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -269,6 +279,25 @@ def proximal_quadratic_perturbation(prox_factory, a, u=None):
+    :math:`a` is the scaling parameter belonging to the quadratic term,
+    :math:`u` is the space element defining the linear functional, and
+    :math:`\\sigma` is the step size.
+
```
Remove redundant newline

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -342,27 +361,48 @@ def proximal_composition(proximal, operator, mu):
+    Given a linear operator :math:`L` with the property that for a scalar
+    :math:`\\mu`
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -342,27 +361,48 @@ def proximal_composition(proximal, operator, mu):
+
+    and a convex function :math:`F`, the following identity holds
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -387,14 +422,26 @@ def proximal_const_func(space):
+    for all values of :math:`x`. The proximal operator of this functional is
+    the identity operator
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -446,6 +483,36 @@ def proximal_box_constraint(space, lower=None, upper=None):
+    If :math:`P` is an interval :math:`[a,b]`, the indicator function is
+    defined as
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -446,6 +483,36 @@ def proximal_box_constraint(space, lower=None, upper=None):
+    For a step size :math:`\\sigma`, the proximal operator of
+    :math:`\\sigma I_{P}` is given by the projection onto the interval
+
+      .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -568,8 +611,37 @@ def proximal_cconj_l2(space, lam=1, g=None):
+
+    The :math:`L_2`-norm/distance :math:`F` is given by is given by
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -568,8 +611,37 @@ def proximal_cconj_l2(space, lam=1, g=None):
+
+    The convex conjugate :math:`F^*` of :math:`F` is given by
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -616,6 +680,25 @@ def proximal_l2(space, lam=1, g=None):
 
+    The :math:`L_2`-norm/distance :math:`F` is given by
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -616,6 +680,25 @@ def proximal_l2(space, lam=1, g=None):
+    For a step size :math:`\\sigma`, the proximal operator of :math:`\\sigma F`
+    is given by
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -705,6 +779,29 @@ def proximal_cconj_l2_squared(space, lam=1, g=None):
+    -----
+    The squared :math:`L_2`-norm/distance :math:`F` is given by
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -705,6 +779,29 @@ def proximal_cconj_l2_squared(space, lam=1, g=None):
+
+    The convex conjugate :math:`F^*` of :math:`F` is given by
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -705,6 +779,29 @@ def proximal_cconj_l2_squared(space, lam=1, g=None):
+    For a step size :math:`\\sigma`, the proximal operator of
+    :math:`\\sigma F^*` is given by
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -775,12 +867,27 @@ def proximal_l2_squared(space, lam=1, g=None):
+    -----
+    The squared :math:`L_2`-norm/distance :math:`F` is given by
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -775,12 +867,27 @@ def proximal_l2_squared(space, lam=1, g=None):
+    For a step size :math:`\\sigma`, the proximal operator of :math:`\\sigma F`
+    is given by
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -847,6 +921,61 @@ def proximal_cconj_l1(space, lam=1, g=None, isotropic=False):
+    -----
+    The :math:`L_1`-norm/distance :math:`F` is the functional
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -847,6 +921,61 @@ def proximal_cconj_l1(space, lam=1, g=None, isotropic=False):
+    The convex conjugate :math:`F^*` of :math:`F` is given by the indicator
+    function of the set :math:`Q_\\lambda`
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -847,6 +921,61 @@ def proximal_cconj_l1(space, lam=1, g=None, isotropic=False):
+    For a step size :math:`\\sigma`, the proximal operator of
+    :math:`\\sigma F^*` is given by
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -847,6 +921,61 @@ def proximal_cconj_l1(space, lam=1, g=None, isotropic=False):
+
+    In this case, the dual is
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -847,6 +921,61 @@ def proximal_cconj_l1(space, lam=1, g=None, isotropic=False):
+    An alternative formulation is available for `ProductSpace`'s, in which
+    case the ``isotropic`` parameter can be used, giving
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -967,11 +1083,43 @@ def proximal_l1(space, lam=1, g=None, isotropic=False):
+    -----
+    The :math:`L_1`-norm/distance :math:`F` is the functional
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -967,11 +1083,43 @@ def proximal_l1(space, lam=1, g=None, isotropic=False):
+    For a step size :math:`\\sigma`, the proximal operator of :math:`\\sigma F`
+    is
+
+        .. math::
```
Too much indentation

adler-j commented on odl/solvers/advanced/proximal_operators.py:
```diff
@@ -967,11 +1083,43 @@ def proximal_l1(space, lam=1, g=None, isotropic=False):
+    An alternative formulation is available for `ProductSpace`'s, where the
+    the ``isotropic`` parameter can be used, giving
+
+        .. math::
```
Too much indentation

adler-j commented:
Great! Look fine to me now, great improvement overall!
